### DISCUSSION TITLE
TreeDataTemplate ItemsSource binding Converter

### DISCUSSION THREAD

**Discussion by PashtetR** [AvaloniaUI/Avalonia#10347] [Q&A]
Hi, Everyone!
I want to save some approaches, porting my WPF application to AvaloniaUI. 

I used binding with converter to set ItemsSource in HierarchicalDataTemplate, according to DataType in WPF.
Something like this:
`HierarchicalDataTemplate DataType="{TagGroup}" ItemsSource="{Binding Converter={StaticResource TreeItemsConverter}}"`

I want apply the same practice. Something like this:
`TreeDataTemplate DataType="{TagGroup}" ItemsSource="{Binding Converter={StaticResource TreeItemsConverter}}"`

TagGroup is my model class.
TreeItemsConverter returns ObservableCollection.

Unfortunately converter isnt called. There is a message in output-window:
`[[Binding]Binding produced invalid value for 'Items' ('System.Collections.IEnumerable'): 'GroupName' ('Project Model.Groups.TagGroup')(TreeViewItem #55094571) `

Other converters with other bindings are working. This converter with TreeView.Items binding (for example) is working too.

Can you help me? Is it a bug or feature? Thanks!

**Comment by timunie** [comment]
Hard to say without seeing the source. Can you show your Converter?

**Comment by PashtetR** [comment]
Converter is trivial just for test:
```
public class TreeItemsConverter : IValueConverter
{
	public object Convert(object? value, Type targetType, object? parameter, CultureInfo culture)
	{
		Debug.WriteLine("Convert TreeItemsConverter");
		return new ObservableCollection<string>() { "test1", "test2" };
	}

	public object ConvertBack(object? value, Type targetType, object? parameter, CultureInfo culture)
	{
		throw new NotImplementedException();
	}
}
```

**Reply by timunie** [reply]
hm that looks okay to me. Any chance you can share a simple test App with us to double-check?

**Reply by PashtetR** [reply]
There is simple App:
https://github.com/PashtetR/AvaloniaApplication1
There are three trees. First - direct binding to ObservableCollections. Second - try to get same ObservableCollections through converter. Third - same converter returns ObservableCollection to Tree.Items and it works.

**Reply by timunie** [reply]
I had a deeper look. Seems to be an issue in Avalonia, where the Converter is ignored internally. This line seems to be the root cause: 

https://github.com/AvaloniaUI/Avalonia/blob/e7e04a067fc8fa7b71ca4256efb81f5cb536ee18/src/Markup/Avalonia.Markup.Xaml/Templates/TreeDataTemplate.cs#L37-L53 

I'll convert this discussion into an issue now. But I don't know how fast it can be solved.

**Reply by PashtetR** [reply]
Ok, thanks for your time and help!
